openmeter/ent/schema/productcatalog.go

```diff
@@ -0,0 +1,251 @@
+package schema
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+
+	"entgo.io/ent"
+	"entgo.io/ent/dialect"
+	"entgo.io/ent/dialect/entsql"
+	"entgo.io/ent/dialect/sql"
+	"entgo.io/ent/schema/edge"
+	"entgo.io/ent/schema/field"
+	"entgo.io/ent/schema/index"
+
+	"github.com/openmeterio/openmeter/openmeter/productcatalog/plan"
+	"github.com/openmeterio/openmeter/pkg/datex"
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+)
+
+type Plan struct {
+	ent.Schema
+}
+
+func (Plan) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.KeyMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (Plan) Fields() []ent.Field {
+	return []ent.Field{
+		field.Int("version").
+			Min(1),
+		field.String("name").
+			NotEmpty(),
+		field.String("description").
+			Optional().
+			Nillable(),
+		field.String("currency").
+			Default("USD").
+			NotEmpty().
+			Immutable(),
+		field.Time("effectiveFrom").
+			Optional().
+			Nillable(),
+		field.Time("effectiveTo").
+			Optional().
+			Nillable(),
+	}
+}
+
+func (Plan) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("phases", PlanPhase.Type).
+			Annotations(entsql.Annotation{
+				OnDelete: entsql.Cascade,
+			}),
+	}
+}
+
+func (Plan) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "key", "version").
+			Unique(),
+	}
+}
+
+type PlanPhase struct {
+	ent.Schema
+}
+
+func (PlanPhase) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.KeyMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (PlanPhase) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("start_after").
+			GoType(datex.ISOString("")).
+			Default(plan.DefaultStartAfter),
+		field.String("name").
+			NotEmpty(),
+		field.String("description").
+			Optional().
+			Nillable(),
+		field.String("discounts").
+			GoType([]plan.Discount{}).
```

Comment:
normalize please, separate PR is ok

Reply:
Hmm. Why do we need the discounts to be normalized here?

Reply:
@hekike please discuss with @tothandras, the two of you are on a different stance regarding normalization so IMO lets come up with an agreement (came up on other PRs as well)